src/integrations/pulsar/consumer.cpp

```diff
@@ -0,0 +1,285 @@
+  std::vector<Message> batch{};
+
+  const auto batch_size = info.batch_size.value_or(kDefaultBatchSize);
+  batch.reserve(batch_size);
+
+  auto remaining_timeout_in_ms = info.batch_interval.value_or(kDefaultBatchInterval).count();
+  auto start = std::chrono::steady_clock::now();
+
+  bool run_batch = true;
+  for (int64_t i = 0; remaining_timeout_in_ms > 0 && i < batch_size && is_running.load(); ++i) {
+    pulsar_client::Message message;
+    const auto result = std::invoke([&] {
+      if constexpr (std::same_as<TConsumer, pulsar_client::Consumer>) {
+        return consumer.receive(message, static_cast<int>(remaining_timeout_in_ms));
+      } else {
+        static_assert(std::same_as<TConsumer, pulsar_client::Reader>);
+        return consumer.readNext(message, static_cast<int>(remaining_timeout_in_ms));
+      }
+    });
+    switch (result) {
+      case pulsar_client::Result::ResultTimeout:
+        run_batch = false;
+        break;
+      case pulsar_client::Result::ResultOk:
+        batch.emplace_back(Message{std::move(message)});
+        break;
+      default:
+        spdlog::warn(fmt::format("Unexpected error while consuming message from consumer {}, error: {}",
+                                 info.consumer_name, result));
+        return {pulsar_client::strResult(result)};
+    }
+
+    if (!run_batch) {
+      break;
+    }
+
+    auto now = std::chrono::steady_clock::now();
+    auto took = std::chrono::duration_cast<std::chrono::milliseconds>(now - start);
+    remaining_timeout_in_ms = remaining_timeout_in_ms - took.count();
+    start = now;
+  }
+
+  return std::move(batch);
+}
+
+class SpdlogLogger : public pulsar_client::Logger {
+  bool isEnabled(Level /*level*/) override { return spdlog::should_log(spdlog::level::trace); }
+
+  void log(Level /*level*/, int /*line*/, const std::string &message) override {
+    spdlog::trace("[Pulsar] {}", message);
+  }
+};
+
+class SpdlogLoggerFactory : public pulsar_client::LoggerFactory {
+  pulsar_client::Logger *getLogger(const std::string & /*file_name*/) override {
+    if (!logger_) {
+      logger_ = std::make_unique<SpdlogLogger>();
+    }
+    return logger_.get();
+  }
+
+ private:
+  std::unique_ptr<SpdlogLogger> logger_;
+};
+
+pulsar_client::Client CreateClient(const std::string &service_url) {
+  static SpdlogLoggerFactory logger_factory;
+  pulsar_client::ClientConfiguration conf;
+  conf.setLogger(&logger_factory);
+  return {service_url, conf};
+}
+}  // namespace
+
+Message::Message(pulsar_client::Message &&message) : message_{std::move(message)} {}
+
+std::span<const char> Message::Payload() const {
+  return {static_cast<const char *>(message_.getData()), message_.getLength()};
+}
+
+Consumer::Consumer(ConsumerInfo info, ConsumerFunction consumer_function)
+    : info_{std::move(info)}, consumer_function_{std::move(consumer_function)} {
+  client_.emplace(CreateClient(info_.service_url));
```

Comment:
It can, but depending on the order you need to use `info_` or `info` because it's moved. To make it explicit I used `std::optional` to delay construction.

Reply:
I have no problem with the order dependent initialization, especially as we have a warning for that in clang `-Wreorder-ctor`. In this case `std::optional` is an additional complexity in my opinion.